DependencyInjection/DoctrineMongoDBExtension.php

```diff
@@ -496,4 +502,71 @@ public function getXsdValidationBasePath()
     {
         return __DIR__ . '/../Resources/config/schema';
     }
+
+    private function registerMetadataCacheWarmer(string $documentManagerName, ContainerBuilder $container): void
+    {
+        $metadataCacheAlias              = $this->getObjectManagerElementName($documentManagerName . '_metadata_cache');
+        $decoratedMetadataCacheServiceId = (string) $container->getAlias($metadataCacheAlias);
+        $phpArrayCacheDecoratorServiceId = $decoratedMetadataCacheServiceId . '.php_array';
+        $phpArrayFile                    = '%kernel.cache_dir%' . sprintf('/doctrine/orm/%s_metadata.php', $documentManagerName);
+        $cacheWarmerServiceId            = $this->getObjectManagerElementName($documentManagerName . '_metadata_cache_warmer');
+
+        $container->register($cacheWarmerServiceId, MetadataCacheWarmer::class)
+            ->setArguments([new Reference(sprintf('doctrine_mongodb.odm.%s_document_manager', $documentManagerName)), $phpArrayFile])
+            ->addTag('kernel.cache_warmer', ['priority' => 1000]); // priority should be higher than ProxyCacheWarmer
+
+        $container->setAlias($metadataCacheAlias, $phpArrayCacheDecoratorServiceId);
+        $container->register($metadataCacheAlias, PhpArrayAdapter::class)
+            ->addArgument($phpArrayFile)
+            ->addArgument(new Definition(ArrayAdapter::class));
+    }
+
+    private function loadMetadataCacheDriver(array $documentManager, ContainerBuilder $container): void
+    {
+        if (! isset($documentManager['metadata_cache_driver'])) {
+            // Automatically configure cache
+            $this->configureDefaultMetadataCache($documentManager, $container);
+
+            return;
+        }
+
+        $this->loadObjectManagerCacheDriver($documentManager, $container, 'metadata_cache');
+
+        // Convert to PSR-6 cache if required
+        $cacheServiceId  = $this->getObjectManagerElementName($documentManager['name'] . '_metadata_cache');
+        $cacheDefinition = $container->getDefinition($cacheServiceId);
+        if (is_a($cacheDefinition->getClass(), CacheItemPoolInterface::class, true)) {
+            return;
+        }
+
+        $adapterDefinition = new Definition(CacheItemPoolInterface::class);
+        $adapterDefinition
+            ->setFactory([CacheAdapter::class, 'wrap'])
+            ->setArgument(0, $cacheDefinition);
+
+        $container->setDefinition($cacheServiceId, $adapterDefinition);
```

Comment:
This part won't work like that, as we can't check the cache definitions class at this stage. In DoctrineBundle, we worked around this by adding a compiler pass that converts caches to the correct instance: https://github.com/doctrine/DoctrineBundle/blob/2.4.x/DependencyInjection/Compiler/CacheCompatibilityPass.php

ODM only has metadata caching, so we don't need to account for other caches like we do for ORM, but the logic can stay the same.

In the extension, we then set up the metadata cache layer the way it has been, with the compiler pass ensuring that the correct cache instance is passed to the configuration.